index.js

```diff
@@ -830,60 +805,108 @@ Tree.prototype._onToggle = function (d) {
+      this.root = _n
+    }
+  }
+
+  return _n
+}
+
 /*
  * Adds a new node to the tree. Pass in d as the data that represents
  * the node, parent (which can be the parent object or an id), and an optional
  * index. If the index is sent, the node will be inserted at that index within the
  * parent's children.
+ *
+ * Receives optional `opts`, which supports:
+ *    `expand`: defaults to true, which will expand the incoming node's parent if the parent is selected
  */
-Tree.prototype.add = function (d, parent, idx) {
+Tree.prototype.add = function (d, parent, idx, opts = { expand: true }) {
   if (this._layout[d.id]) {
-    // can't add a node that we already have
+    // Already have this node, ignore
     return
   }
 
-  // internal node used for computing the layout
-  var _d = { id: d.id }
-
-  if (!parent && this.options.forest) {
-    // Forest tree and the new node is a new root
-    this.nodes[d.id] = d // Store the real node
-    this._layout[_d.id] = _d
-    this.root.splice(typeof idx === 'number' ? idx : this.root.length, 0, _d)
-    this._transitionWrap(this._slide)()
-    return d
-  } else if (parent) {
-    parent = this._layout[parent !== null && typeof parent === 'object' ? parent.id : parent]
-  } else if (!parent && !this.root) {
-    this.root = _d
-  } else {
-    // No parent, and not a new root node
-    return
-  }
+  // Turn parent into a parent object
+  parent = this._layout[parent !== null && typeof parent === 'object' ? parent.id : parent]
 
-  if (parent && parent.selected && parent.collapsed) {
+  if (opts.expand && parent && parent.selected && parent.collapsed) {
     // The parent is selected and collapsed, we want to expand its children (#259)
     delete parent.collapsed
     // Show the children immediately
     this._fly()
     this._forceRedraw()
   }
 
-  _d.parent = parent
-  this.nodes[d.id] = d
-  this._layout[_d.id] = _d
+  let _d = this._store(d, parent, idx)
+
+  this._transitionWrap(this._slide)()
 
-  if (typeof idx !== 'undefined') {
-    parent._allChildren.splice(idx, 0, _d)
-  } else if (parent) {
-    if (!parent._allChildren) {
-      parent._allChildren = []
+  return d
+}
+
+/*
+ * Adds multiple nodes to the tree. This is more efficient than invoking `.add` on each node
+ * as it only draws the nodes once, after the internal tree representation has been updated.
+ * Each node in the array should contain the arguments that should be passed to `.add`:
+ *  ```
+ *  {
+ *    data: // the node's data
+ *    parent: // optional parent of the node
+ *    idx: // idx to insert in the parent
+ *  }
+ *  ```
+ */
+Tree.prototype.addAll = function (nodes = []) {
+  nodes.forEach(node => {
+    if (this._layout[node.data.id]) {
+      // ignore this node. already in the tree
+      return
     }
-    parent._allChildren.push(_d)
-  }
+    // Turn parent into a parent object
+    let parent = this._layout[node.parent !== null && typeof node.parent === 'object' ? node.parent.id : node.parent]
```

Comment:
Could this line be moved to `_store` to further reduce duplication (something similar is in the `add` method)

Reply:
Probably. I'll check.